File: csharp/DatesTest.cs
using System;
using Microsoft.VisualStudio.TestTools.UnitTesting;

namespace csharp
{
    [TestClass]
    public class DatesTest
    {
        private static DateTime Now = DateTime.Now;
        private static DateTime SevenDaysAgo = Now.AddDays(-7);
        private static DateTime SevenDaysFuture = Now.AddDays(7);
        private static DateTime FourteenDaysFuture = Now.AddDays(14);
        private static Overlapper Overlapper = new Overlapper();

        [TestMethod]
        public void TestRangeOneExtendsIntoRangeTwo()
        {
            var dateRange1 = new DateRange(SevenDaysAgo, SevenDaysFuture);
            var dateRange2 = new DateRange(Now, FourteenDaysFuture);
            var expected = new DateRange(Now, SevenDaysFuture);
            var actual = Overlapper.FindOverlap(dateRange1, dateRange2);
            Assert.IsTrue(actual.Equals(expected));
        }

        [TestMethod]
        public void TestRangeTwoExtendsIntoRangeOne()
        {
            var dateRange2 = new DateRange(SevenDaysAgo, SevenDaysFuture);
            var dateRange1 = new DateRange(Now, FourteenDaysFuture);
            var expected = new DateRange(Now, SevenDaysFuture);
            var actual = Overlapper.FindOverlap(dateRange1, dateRange2);
            Assert.IsTrue(actual.Equals(expected));
        }

        [TestMethod]
        [ExpectedException(typeof(Exception))]
        public void TestRangeOneEndsWhenRangeTwoBegins()
        {
            var dateRange1 = new DateRange(SevenDaysAgo, Now);
            var dateRange2 = new DateRange(Now, FourteenDaysFuture);
            Overlapper.FindOverlap(dateRange1, dateRange2);
        }

        [TestMethod]
        [ExpectedException(typeof(Exception))]
        public void TestRangeTwoEndsWhenRangeOneBegins()
        {
            var dateRange2 = new DateRange(SevenDaysAgo, Now);
            var dateRange1 = new DateRange(Now, FourteenDaysFuture);
            Overlapper.FindOverlap(dateRange1, dateRange2);
        }

        [TestMethod]
        public void TestRangeOneContainsRangeTwo()
        {
            var dateRange1 = new DateRange(SevenDaysAgo, FourteenDaysFuture);
            var dateRange2 = new DateRange(Now, SevenDaysFuture);
            var expected = new DateRange(Now, SevenDaysFuture);
            var actual = Overlapper.FindOverlap(dateRange1, dateRange2);
            Assert.IsTrue(actual.Equals(expected));
        }

        [TestMethod]
        public void TestRangeTwoContainsRangeOne()
        {
            var dateRange2 = new DateRange(SevenDaysAgo, FourteenDaysFuture);
            var dateRange1 = new DateRange(Now, SevenDaysFuture);
            var expected = new DateRange(Now, SevenDaysFuture);
            var actual = Overlapper.FindOverlap(dateRange1, dateRange2);
            Assert.IsTrue(actual.Equals(expected));
        }

        [TestMethod]
        [ExpectedException(typeof(Exception))]
        public void TestRangeOneDoesNotOverlapWithRangeTwo()
        {
            var dateRange1 = new DateRange(SevenDaysAgo, Now);
            var dateRange2 = new DateRange(SevenDaysFuture, FourteenDaysFuture);
            Overlapper.FindOverlap(dateRange1, dateRange2);
        }

        [TestMethod]
        [ExpectedException(typeof(Exception))]
        public void TestRangeTwoDoesNotOverlapWithRangeOne()
        {
            var dateRange2 = new DateRange(SevenDaysAgo, Now);
            var dateRange1 = new DateRange(SevenDaysFuture, FourteenDaysFuture);
            Overlapper.FindOverlap(dateRange1, dateRange2);
        }

        [TestMethod]
        public void TestRangeOneEqualsRangeTwo()
        {
            var dateRange1 = new DateRange(Now, SevenDaysFuture);
            var dateRange2 = new DateRange(Now, SevenDaysFuture);
            var expected = new DateRange(Now, SevenDaysFuture);
            var actual = Overlapper.FindOverlap(dateRange1, dateRange2);
            Assert.IsTrue(actual.Equals(expected));
        }
    }
}

File: py/dates_test.py
import unittest
import datetime

import dates

class TestStringMethods(unittest.TestCase):
    def setUp(self):
        self.now = datetime.datetime.now()
        self.seven_days_ago = self.now - datetime.timedelta(days=7)
        self.seven_days_future = self.now + datetime.timedelta(days=7)
        self.fourteen_days_future = self.now + datetime.timedelta(days=14)

    def test_range_one_extends_into_range_two(self):
        dateRange1 = dates.DateRange(self.seven_days_ago, self.seven_days_future)
        dateRange2 = dates.DateRange(self.now, self.fourteen_days_future)
        expected = dates.DateRange(self.now, self.seven_days_future)
        actual = dates.find_overlap(dateRange1, dateRange2)
        
        self.assertEqual(expected.begin, actual.begin)
        self.assertEqual(expected.end, actual.end)

    def test_range_two_extends_into_range_one(self):
        dateRange2 = dates.DateRange(self.seven_days_ago, self.seven_days_future)
        dateRange1 = dates.DateRange(self.now, self.fourteen_days_future)
        expected = dates.DateRange(self.now, self.seven_days_future)
        actual = dates.find_overlap(dateRange1, dateRange2)
        
        self.assertEqual(expected.begin, actual.begin)
        self.assertEqual(expected.end, actual.end)

    def test_range_one_ends_at_range_two_begin(self):
        dateRange1 = dates.DateRange(self.seven_days_ago, self.now)
        dateRange2 = dates.DateRange(self.now, self.fourteen_days_future)
        self.assertRaises(Exception, dates.find_overlap, dateRange1, dateRange2)

    def test_range_two_ends_at_range_one_begin(self):
        dateRange2 = dates.DateRange(self.seven_days_ago, self.now)
        dateRange1 = dates.DateRange(self.now, self.fourteen_days_future)
        self.assertRaises(Exception, dates.find_overlap, dateRange1, dateRange2)

    def test_range_one_contains_range_two(self):
        dateRange1 = dates.DateRange(self.seven_days_ago, self.fourteen_days_future)
        dateRange2 = dates.DateRange(self.now, self.seven_days_future)
        expected = dates.DateRange(self.now, self.seven_days_future)
        actual = dates.find_overlap(dateRange1, dateRange2)
        
        self.assertEqual(expected.begin, actual.begin)
        self.assertEqual(expected.end, actual.end)

    def test_range_two_contains_range_one(self):
        dateRange2 = dates.DateRange(self.seven_days_ago, self.fourteen_days_future)
        dateRange1 = dates.DateRange(self.now, self.seven_days_future)
        expected = dates.DateRange(self.now, self.seven_days_future)
        actual = dates.find_overlap(dateRange1, dateRange2)
        
        self.assertEqual(expected.begin, actual.begin)
        self.assertEqual(expected.end, actual.end)

    def test_range_one_does_not_overlap_with_range_two(self):
        dateRange1 = dates.DateRange(self.seven_days_ago, self.now)
        dateRange2 = dates.DateRange(self.seven_days_future, self.fourteen_days_future)
        self.assertRaises(Exception, dates.find_overlap, dateRange1, dateRange2)

    def test_range_two_does_not_overlap_with_range_one(self):
        dateRange2 = dates.DateRange(self.seven_days_ago, self.now)
        dateRange1 = dates.DateRange(self.seven_days_future, self.fourteen_days_future)
        self.assertRaises(Exception, dates.find_overlap, dateRange1, dateRange2)

    def test_range_one_equals_range_two(self):
        dateRange2 = dates.DateRange(self.now, self.seven_days_future)
        dateRange1 = dates.DateRange(self.now, self.seven_days_future)
        expected = dates.DateRange(self.now, self.seven_days_future)
        actual = dates.find_overlap(dateRange1, dateRange2)
        
        self.assertEqual(expected.begin, actual.begin)
        self.assertEqual(expected.end, actual.end)

if __name__ == '__main__':
    unittest.main()